utilities/transactions/write_prepared_txn_db.h

```diff
@@ -401,30 +401,56 @@ class WritePreparedCommitEntryPreReleaseCallback : public PreReleaseCallback {
  public:
   // includes_data indicates that the commit also writes non-empty
   // CommitTimeWriteBatch to memtable, which needs to be committed separately.
-  WritePreparedCommitEntryPreReleaseCallback(
-      WritePreparedTxnDB* db, DBImpl* db_impl,
-      SequenceNumber prep_seq = kMaxSequenceNumber, bool includes_data = false)
+  WritePreparedCommitEntryPreReleaseCallback(WritePreparedTxnDB* db,
+                                             DBImpl* db_impl,
+                                             SequenceNumber prep_seq,
+                                             size_t prep_batch_cnt,
+                                             size_t data_batch_cnt = 0)
       : db_(db),
         db_impl_(db_impl),
         prep_seq_(prep_seq),
-        includes_data_(includes_data) {}
+        prep_batch_cnt_(prep_batch_cnt),
+        data_batch_cnt_(data_batch_cnt),
+        includes_data_(data_batch_cnt_ > 0) {
+    assert((prep_batch_cnt_ > 0) != (prep_seq == kMaxSequenceNumber));  // xor
+    assert(prep_batch_cnt_ > 0 || data_batch_cnt_ > 0);
+  }
 
   virtual Status Callback(SequenceNumber commit_seq) {
     assert(includes_data_ || prep_seq_ != kMaxSequenceNumber);
+    const uint64_t last_commit_seq = LIKELY(data_batch_cnt_ <= 1)
+                                         ? commit_seq
+                                         : commit_seq + data_batch_cnt_ - 1;
     if (prep_seq_ != kMaxSequenceNumber) {
-      db_->AddCommitted(prep_seq_, commit_seq);
+      if (LIKELY(prep_batch_cnt_ == 1)) {
```

Comment:
why you need the if-else block? Compiler will likely to expand the loop for you (at least with -O3).